rust/agents/relayer/src/msg/op_queue.rs

```diff
@@ -0,0 +1,305 @@
+
+pub type QueueOperation = Box<dyn PendingOperation>;
+
+/// Queue of generic operations that can be submitted to a destination chain.
+/// Includes logic for maintaining queue metrics by the destination and `app_context` of an operation
+#[derive(Debug, Clone, new)]
+pub struct OpQueue {
+    metrics: IntGaugeVec,
+    queue_metrics_label: String,
+    retry_rx: MpmcReceiver<MessageRetryRequest>,
+    #[new(default)]
+    queue: Arc<Mutex<BinaryHeap<Reverse<QueueOperation>>>>,
+}
+
+impl OpQueue {
+    /// Push an element onto the queue and update metrics
+    pub async fn push(&self, op: QueueOperation) {
+        // increment the metric before pushing onto the queue, because we lose ownership afterwards
+        self.get_operation_metric(op.as_ref()).inc();
+
+        self.queue.lock().await.push(Reverse(op));
+    }
+
+    /// Pop an element from the queue and update metrics
+    pub async fn pop(&mut self) -> Option<Reverse<QueueOperation>> {
+        self.process_retry_requests().await;
+        let op = self.queue.lock().await.pop();
+        op.map(|op| {
+            // even if the metric is decremented here, the operation may fail to process and be re-added to the queue.
+            // in those cases, the queue length will decrease to zero until the operation is re-added.
+            self.get_operation_metric(op.0.as_ref()).dec();
+            op
+        })
+    }
+
+    pub async fn process_retry_requests(&mut self) {
+        // TODO: could rate-limit ourselves here, but we expect the volume of messages over this channel to
+        // be very low.
+        // The other consideration is whether to put the channel receiver in the OpQueue or in a dedicated task
+        // that also holds an Arc to the Mutex. For simplicity, we'll put it in the OpQueue for now.
+        let mut message_retry_requests = vec![];
+        while let Ok(message_id) = self.retry_rx.receiver.try_recv() {
+            message_retry_requests.push(message_id);
+        }
+        if message_retry_requests.is_empty() {
+            return;
+        }
+        let mut queue = self.queue.lock().await;
+        let mut repriotized_queue: BinaryHeap<_> = queue
+            .drain()
+            .map(|Reverse(mut e)| {
+                // Can check for equality here because of the PartialEq implementation for MessageRetryRequest,
+                // but can't use `contains` because the types are different
+                if message_retry_requests.iter().any(|r| r == e) {
+                    let hex_id = e.id().encode_hex::<String>();
+                    let destination_domain = e.destination_domain().to_string();
+                    info!(
+                        id = hex_id,
+                        destination_domain, "Retrying OpQueue operation"
+                    );
+                    e.reset_attempts()
+                }
+                Reverse(e)
+            })
+            .collect();
+        queue.append(&mut repriotized_queue);
```

Comment:
we can't assign to queue directly because it's technically wrapped in a mutex. I think the reason we're able to call `mut` methods on it is deref corecion. 

I agree it would've been more efficient to just swap pointers instead of overwriting the heap, but this function is already O(N) so nbd

Reply:
ah oops sorry, makes sense, forgot about the mutex